### PR TITLE
DRILL-3714: Query runs out of memory and remains in CANCELLATION_REQU…

### PR DESCRIPTION
**Pull request by adeneche**
…ESTED state until drillbit is restarted

RpcBus.ChannelClosedHandler calls CoordinationQueue.channelClosed() on both server and client side

### CODE REVIEW

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java [hunk 1]
```diff
@@ -159,19 +159,15 @@ public ChannelClosedHandler(C clientConnection, Channel channel) {
...
-      if (RpcBus.this.isClient()) {
-        if(local != null) {
-          logger.info(String.format(msg));
-        }
-      } else {
-        queue.channelClosed(new ChannelClosedException(msg));
-      }
+      logger.info(msg); // should we leave this at info level ?
+
+      queue.channelClosed(new ChannelClosedException(msg));
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java]: @adeneche @jacques-n correct me if I am wrong.

Per my understanding, this logic is incomplete, with or without this change. Let's looks at bit-to-bit comm.

There is one CoordinationQueue **for each instance** of RpcBus (\*Server, \*Client classes inherit from RpcBus). Also, this queue is used by requestors to listen to outcomes of requests.

1. DataClient <--> DataServer. DataClient is always the requestor, and there can be at most two data connections between two bits (A --> B and B --> A).
  a. Since a DataClient is created per connection, the client's queue contains outcomes of requests to one DataServer. When this connection closes, failing all RPC outcomes in the queue makes sense.
  b. There is only one instance of DataServer per Drillbit, and so **one server queue**. Since DataServer never makes requests, this queue should be empty. `queue.channelClosed(...)` should be a noop.

2. ControlClient <--> ControlServer. This communication is peer-to-peer i.e. ControlServer and ControlClient can make requests and handle requests. The bit initiating a connection is the ControlClient, and its peer is the ControlServer for lifetime of this connection. There can be at most one connection between two bits (A <--> B), and messages are sent both ways. Now, assume a connection is made.
  a. On the ControlClient side, the queue contains outcomes of requests to one ControlServer. When this connection closes, failing all RPC outcomes in the queue makes sense.
  b. There is only one instance of ControlServer per Drillbit, and so **one server queue**. However, ControlServer can make requests to other bits, and to multiple clients! So this queue can contain outcomes from multiple connections and `queue.channelClosed(...)` fails outcomes of requests from **all** connections??

Comment (jacques-n) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java]: Nice catch Sudheesh.

Reflecting on this now, I think there is a much cleaner way to structure this than the current implementation. If we manage this at the connection level (new "outstanding handler => rpc id mapping" per connection), we can rely on Netty's RPC thread properties to have a trivial structure that doesn't need to be thread safe and is closed out on connection termination. Do one of you want to pick up that modification? I think that would make things much nicer.

Comment (adeneche) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java]: @sudheeshkatkam this is also true for the user tunnel. We have a single UserServer on each drillbit that has connections with multiple clients, if one connection is closed, the queue will fail all batches waiting for outcome. See [DRILL-3763](https://issues.apache.org/jira/browse/DRILL-3763)

@jacques-n let me wrap my head around your suggestion. In the meantime, I may have a simple fix that would resolve the issue in the short-term: we just need to pass the closed channel to the coordination queue and only fail the listeners related to the closed channel.

Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java]: Makes sense that this problem shows up in UserServer <--> UserClient comm.; not just [DRILL-3763](https://issues.apache.org/jira/browse/DRILL-3763), I think [DRILL-3823](https://issues.apache.org/jira/browse/DRILL-3823), [DRILL-3833](https://issues.apache.org/jira/browse/DRILL-3833), [DRILL-4170](https://issues.apache.org/jira/browse/DRILL-4170) and [DRILL-4298](https://issues.apache.org/jira/browse/DRILL-4298) will also be resolved :)

Comment (adeneche) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java]: That would be a highly valuable fix then :D

DRILL-3823 seems unrelated though, maybe a typo in the JIRA number ?

**Comment by adeneche** [comment]
Just updated the fix to only fail the CoordinationQueue listeners for the closed channel. @jacques-n and @sudheeshkatkam let me know if this is acceptable. In the meantime, I'll see if I can add a unit test for [DRILL-3763](https://issues.apache.org/jira/browse/DRILL-3763)

**Comment by adeneche** [comment]
I'm seeing functional failures for the updated fix. Investigating those

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java (hunk 1 above)
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java]: I meant [DRILL-3826](https://issues.apache.org/jira/browse/DRILL-3826).

Looks like you noted that [DRILL-3241](https://issues.apache.org/jira/browse/DRILL-3241) is also related.

**Comment by sudheeshkatkam** [comment]
+0 Why do we need a work around?

I think Jacques is suggesting to move the logic to RemoteConnection, which is much cleaner.

**Comment by jacques-n** [comment]
I'm working on an alternative approach where the coordination queue is associated with each connection. Upon review, it still needs thread protection because you have contention between sending threads and the socket thread but it can be much simpler/cleaner. I'm still getting some test failures but hope to have a proposal up today.

**Comment by adeneche** [comment]
Well the workaround is already implemented and as I stated in the JIRA, my knowledge around Netty is basic so it will probably take me quite some time to figure out a cleaner solution.

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java [hunk 2]
```diff
@@ -159,19 +159,15 @@ public ChannelClosedHandler(C clientConnection, Channel channel) {
...
-          logger.info(String.format(msg));
-        }
-      } else {
-        queue.channelClosed(new ChannelClosedException(msg));
-      }
+      logger.info(msg); // should we leave this at info level ?
+
+      queue.channelClosed(new ChannelClosedException(msg), future.channel());
 
       clientConnection.close();
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java]: Slightly unrelated: is this `clientConnection.close();` call unnecessary, since this listener is meant to handle the underlying channel closure?

Comment (jacques-n) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java]: I believe this is still required because we could be in a backpressure blocking state and need to error out the pending messages. My patch tries to consolidate the connection close behavior (so we have one place to do work instead of two). Note that right now we add close handlers in both the connection and the rpc bus which is confusing/unclean.

**Comment by jacques-n** [comment]
I came up with an alternative approach as I mentioned above. Basically moving coordination management to the RemoteConnection instead of handling at the RpcBus level.

https://github.com/apache/drill/pull/463
